docs/development/codeguide.rst

```diff
@@ -203,8 +203,9 @@ Coding Style/Conventions
     import numpy as np
     import matplotlib as mpl
     import matplotlib.pyplot as plt
+    import pandas as pd
```

Comment:
If anybody ever deviates from the naming conventions ICN001 enforces then the Ruff hook in `pre-commit` will clearly state where in the code the violation occurred and what name it expects to be used. It is even possible to have Ruff fix the violations automatically. I don't think this text in the developer documentation is informative anymore.

Reply:
Personally I prefer to know what I am expected to write before a pre-commit tool tells me I've done it wrong and how to fix it. Besides, we are still seeing issues with contributors who are unable to set up their `pre-commit` locally.
If we have section in the docs stating in some detail what import styles Ruff is or is going to enforce via `isort`, `flake8` conventions and what else, that might replace this explicit list, but currently I don't see such a reference. Opinions may vary on this.

Reply:
I'm not sure if people are actually having problems installing `pre-commit`. I consider it much more likely that they haven't followed our instructions to install `pre-commit` simply because they have not read our developer documentation at all, in which case nothing we write there is going to help them.

I'd rather not write down any specific rules that Ruff would be enforcing anyways because I believe the shorter our developer documentation is the more people might actually read it.

Reply:
I suspect similarly. I've installed pre-commit on everything from a Python env on my phone (bad idea, don't do dev work on the phone) to a compute cluster. I've never once run into a problem with the pip install except when I didn't have git on a machine.

Reply:
Currently the docs do not even state that you must install and activate pre-commit before you are allowed to push anything upstream, so that would need to be changed along the way. And then how would you handle cases when a series of commits fail in the CI pre-commit, you ask the contributor to run pre-commit locally and they state they have it installed, yet it is apparently allowing invalid commits? I thought the point of all the pre-commit tools was to lower the threshold for new contributors, but that is not always reflected by my experience – apparently it is not that easy for everyone, or are those just some freak cases of exotic Windows installations?

Reply:
> Currently the docs do not even state that you must install and activate pre-commit before you are allowed to push anything upstream, so that would need to be changed along the way.

In our documentation we indeed only say that we _encourage_ installing `pre-commit`, and I do think we should change that.

> And then how would you handle cases when a series of commits fail in the CI pre-commit, you ask the contributor to run pre-commit locally and they state they have it installed, yet it is apparently allowing invalid commits?

In such a case we can conclude that they have not installed our `pre-commit` hooks. In my usage `pre-commit` has proven itself to be trustworthy. A first-time contributor who needs multiple commits to get the code style checks to pass has proven themselves to be untrustworthy when it comes to code style and the related tooling.

> I thought the point of all the pre-commit tools was to lower the threshold for new contributors, but that is not always reflected by my experience...

`pre-commit` can only help those who use it.

> ... apparently it is not that easy for everyone, or are those just some freak cases of exotic Windows installations?

@pllim, do you know how simple or complicated it is to install `pre-commit` on Windows?

Updating our developer documentation properly is beyond the scope of this pull request, so I agree with merging this as is, assuming that the commits get squashed.

Reply:
Re: Windows -- I think I used both conda and pip on different times and they both worked, but I mostly work on WSL2.

Reply:
Is this conversation blocking merge? Should I wait?

Reply:
>  A first-time contributor who needs multiple commits to get the code style checks to pass has proven themselves to be untrustworthy when it comes to code style and the related tooling.

It is technically possible to get multiple commits to pass and then fail pre-commit. But this requires rebases and other fancy stuff that then skip the "pre-commit".
Our CI run of pre-commit should tell any contributor exactly what to fix. And the comment `pre-commit autofix` can take care of many stuff server side.
Installing pre-commit locally AFAIK is always as easy as `pip install pre-commit; cd <git repo>; pre-commit install` and makes their life much easier.

Reply:
> Is this conversation blocking merge? Should I wait?

I think we should update some of our developer documentation, but doing that properly is beyond the scope here and shouldn't block merging.

Reply:
Indeed, this is moving to an entirely different topic and certainly needs more discussion.
The PR was already squashed; I probably would not have added the `as np` imports just to rename them later, but I think our commit history can live with that. ;-)

Reply:
To conclude this part of the discussion here, in my experience it simply is not straightforward for every first-time contributor to parse a report like https://results.pre-commit.ci/run/github/2081289/1703509121.Z6WkqI4KQVm9d2te2s3qhg and make the required fixes, and if I see such commits coming in even after mentioning the use of `pre-commit`, I am still hesitant to label anyone as "untrustworthy" without knowing myself about any possible pitfalls on a given platform. For which we don't seem to be able to provide that much guidance e.g. for the various Windows platforms.